streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamKStreamJoin.java

```diff
@@ -228,7 +231,7 @@ private void emitNonJoinedOuterRecords(
                     sharedTimeTracker.minTime = timestamp;
 
                     // Skip next records if window has not closed
-                    if (timestamp + joinAfterMs + joinGraceMs >= sharedTimeTracker.streamTime) {
+                    if (sharedTimeTracker.minTime + windowsAfterIntervalMs  + joinGraceMs >= sharedTimeTracker.streamTime) {
```

Comment:
I'm not sure about this part. If the `value` has right value and this is on right side, why don't we check with `windows.beforeMs` to see if it expires? Not sure why `windowsAfterIntervalMs` is always set to `windows.afterMs`.